# HW5 Notebook
Name: **Tony Nguyen**

Class: CPSC 322 01

Term: Fall 2023

## 1. Load libraries and datasets

### 1.1 Import the data table and utility functions.

In [1]:
from data_table import *
from data_learn import *
from data_eval import *
from data_util import *

### 1.2 Load and clean auto data.

In [2]:
auto = DataTable(['mpg','cyls','disp','hp','weight','accl','year','origin','name'])
auto.load('auto-mpg.txt')
auto.rows(range(10))

  mpg    cyls    disp    hp    weight    accl    year    origin  name
-----  ------  ------  ----  --------  ------  ------  --------  -------------------------
   18       8     307   130      3504    12        70         1  chevrolet chevelle malibu
   15       8     350   165      3693    11.5      70         1  buick skylark 320
   18       8     318   150      3436    11        70         1  plymouth satellite
   16       8     304   150      3433    12        70         1  amc rebel sst
   17       8     302   140      3449    10.5      70         1  ford torino
   15       8     429   198      4341    10        70         1  ford galaxie 500
   14       8     454   220      4354     9        70         1  chevrolet impala
   14       8     440   215      4312     8.5      70         1  plymouth fury iii
   14       8     455   225      4425    10        70         1  pontiac catalina
   15       8     390   190      3850     8.5      70         1  amc ambassador dpl

In [3]:
print("Number of rows before cleaning:", auto.row_count())

Number of rows before cleaning: 317


### 1.3 Remove duplicates rows

In [4]:
auto = remove_duplicates(auto)
print("Number of rows after removing duplicated rows:", auto.row_count())

Number of rows after removing duplicated rows: 316


### 1.3 Remove missing rows

In [5]:
auto = remove_missing(auto, auto.columns())
print("Number of rows after removing missing values:", auto.row_count())

Number of rows after removing missing values: 307


## 2. Exploring k-NN

### 2.1 Discretize MPG values

In [6]:
auto_copy = auto.copy()
mpg_value = auto_copy.get_column_data('mpg')
mpg_value = sorted(mpg_value)
bin_width = (mpg_value[-1] - mpg_value[0]) / 3
cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width]
discretize(auto_copy, 'mpg', cut_points)

In [7]:
auto_copy.rows(range(20))

  mpg    cyls    disp    hp    weight    accl    year    origin  name
-----  ------  ------  ----  --------  ------  ------  --------  ----------------------------
    1       8     307   130      3504    12        70         1  chevrolet chevelle malibu
    1       8     350   165      3693    11.5      70         1  buick skylark 320
    1       8     318   150      3436    11        70         1  plymouth satellite
    1       8     304   150      3433    12        70         1  amc rebel sst
    1       8     302   140      3449    10.5      70         1  ford torino
    1       8     429   198      4341    10        70         1  ford galaxie 500
    1       8     454   220      4354     9        70         1  chevrolet impala
    1       8     440   215      4312     8.5      70         1  plymouth fury iii
    1       8     455   225      4425    10        70         1  pontiac catalina
    1       8     390   190      3850     8.5      70         1  amc ambassador dpl
    1    

### 2.2 Normalize all columns except model and origin

In [8]:
normalize(auto_copy, 'cyls')
normalize(auto_copy, 'disp')
normalize(auto_copy, 'hp')
normalize(auto_copy, 'weight')
normalize(auto_copy, 'accl')

In [9]:
auto_copy.rows(range(20))

  mpg    cyls       disp        hp    weight       accl    year    origin  name
-----  ------  ---------  --------  --------  ---------  ------  --------  ----------------------------
    1     1    0.617571   0.456522  0.53615   0.238095       70         1  chevrolet chevelle malibu
    1     1    0.728682   0.646739  0.589736  0.208333       70         1  buick skylark 320
    1     1    0.645995   0.565217  0.51687   0.178571       70         1  plymouth satellite
    1     1    0.609819   0.565217  0.516019  0.238095       70         1  amc rebel sst
    1     1    0.604651   0.51087   0.520556  0.14881        70         1  ford torino
    1     1    0.932817   0.826087  0.773462  0.119048       70         1  ford galaxie 500
    1     1    0.997416   0.945652  0.777148  0.0595238      70         1  chevrolet impala
    1     1    0.96124    0.918478  0.76524   0.0297619      70         1  plymouth fury iii
    1     1    1          0.972826  0.797278  0.119048       70         1  

### 2.3 Create train and test sets using Holdout

In [10]:
train, test = holdout(auto_copy, auto_copy.row_count()//2)

In [11]:
train.rows(range(10))

  mpg    cyls      disp        hp    weight       accl    year    origin  name
-----  ------  --------  --------  --------  ---------  ------  --------  -----------------------
    1       1  0.645995  0.565217  0.51687   0.178571       70         1  plymouth satellite
    1       1  0.604651  0.51087   0.520556  0.14881        70         1  ford torino
    1       1  0.932817  0.826087  0.773462  0.119048       70         1  ford galaxie 500
    1       1  0.997416  0.945652  0.777148  0.0595238      70         1  chevrolet impala
    1       1  1         0.972826  0.797278  0.119048       70         1  pontiac catalina
    1       1  0.832041  0.782609  0.63425   0.0297619      70         1  amc ambassador dpl
    1       1  0.813953  0.673913  0.552878  0.119048       70         1  dodge challenger se
    1       1  0.702842  0.619565  0.56592   0              70         1  plymouth 'cuda 340
    1       1  0.857881  0.565217  0.609016  0.0892857      70         1  chevrolet monte c

In [12]:
test.rows(range(10))

  mpg    cyls        disp        hp     weight      accl    year    origin  name
-----  ------  ----------  --------  ---------  --------  ------  --------  -----------------------
    3     0.2  0.0594315   0.076087  0.0530196  0.5           78         3  honda civic cvcc
    1     1    0.857881    0.782609  0.768925   0.25          77         1  chrysler cordoba
    1     1    0.728682    0.592391  0.778849   0.410714      79         1  buick estate wagon (sw)
    3     0.2  0.00775194  0.103261  0.0632265  0.77381       74         3  toyota corolla 1200
    2     0.2  0.186047    0.141304  0.223419   0.684524      73         1  chevrolet vega
    1     0.6  0.470284    0.173913  0.555997   0.77381       76         1  ford granada ghia
    1     0.6  0.338501    0.277174  0.329175   0.446429      70         1  amc hornet
    2     0.2  0.136951    0.375     0.299972   0.327381      75         2  saab 99le
    2     0.2  0.22739     0.320652  0.320953   0.517857      78         1  ply

### 2.4 Run k-NN over the train and test set
* k = 5
* Majority vote
* Numerical columns: cylinders, weight, and acceleration
* No nominal attributes

In [13]:
result = knn_eval(train, test, majority_vote, 5, 'mpg', ['cyls','weight','accl'], [])

### 2.5 Print the Confusion Matrix

In [14]:
print(result)

actual      1    2    3    total
--------  ---  ---  ---  -------
1          60   15    2       77
2           6   44   13       63
3           0    4    9       13
total      66   63   24      153


### 2.6 Calculate the Accuracy & Macro average f-measure for MPG

Find the number of non-empty classes (L)

In [15]:
sum_of_actual_labels = result.get_column_data('total')
non_empty_label = 0
for i in sum_of_actual_labels:
    if i != 0:
        non_empty_label += 1

if result[result.row_count() - 1]['actual'] == 'total':
    non_empty_label -= 1

print("Non-empty labels count:", non_empty_label)

Non-empty labels count: 3


Find the average accuracy

In [16]:
total_accuracy = 0
for i in range(1, 4): # labels are 1, 2, 3
    num = accuracy(result, i)
    if num != -1:
        total_accuracy += num

print("Average accuracy:", total_accuracy / non_empty_label)

Average accuracy: 0.8257080610021786


Find macro-average f-measure

In [17]:
total_precision = 0
denum = non_empty_label
for i in range(1, 4): # labels are 1, 2, 3
    num = precision(result, i)
    if num != -1:
        total_precision += num
    else:
        print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
        denum -= 1

precision_m = total_precision / denum
print("Precision:", precision_m)

Precision: 0.6608345358345358


In [18]:
total_recall = 0
denum = non_empty_label
for i in range(1, 4): # labels are 1, 2, 3
    num = recall(result, i)
    if num != -1:
        total_recall += num
    else:
        print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
        denum -= 1

recall_m = total_recall / denum
print("Recall:", recall_m)

Recall: 0.7233137233137233


In [19]:
f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
print("F-score:", f_score)

F-score: 0.6906640028618143


### 2.7 Average result after running three times using Majority Vote

In [20]:
total_accuracy_after_iteration = 0
total_precision_after_iteration = 0
total_recall_after_iteration = 0
total_f_score_after_iteration = 0

for i in range(3):
    # Discretize
    auto_copy = auto.copy()
    mpg_value = auto_copy.get_column_data('mpg')
    mpg_value = sorted(mpg_value)
    bin_width = (mpg_value[-1] - mpg_value[0]) / 3
    cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width]
    discretize(auto_copy, 'mpg', cut_points)

    # Normalize
    normalize(auto_copy, 'cyls')
    normalize(auto_copy, 'disp')
    normalize(auto_copy, 'hp')
    normalize(auto_copy, 'weight')
    normalize(auto_copy, 'accl')

    # Holdout
    train, test = holdout(auto_copy, auto_copy.row_count()//2)

    # kNN
    result = knn_eval(train, test, majority_vote, 5, 'mpg', ['cyls','weight','accl'], [])
    label_list = result.get_column_data('actual')
    if 'total' in label_list:
        label_list.remove('total')

    print("Iteration", i+1)
    print("--------------------")
    print(result)
    print()

    # Find non-empty labels
    sum_of_actual_labels = result.get_column_data('total')
    non_empty_label = 0
    for i in sum_of_actual_labels:
        if i != 0:
            non_empty_label += 1

    if result[result.row_count() - 1]['actual'] == 'total':
        non_empty_label -= 1

    print("Non-empty labels count:", non_empty_label)

    # Accuracy
    total_accuracy = 0
    for i in label_list: 
        num = accuracy(result, i)
        if num != -1:
            total_accuracy += num
    acc = total_accuracy / non_empty_label
    total_accuracy_after_iteration += acc
    print("Average accuracy:", acc)

    # Precision
    total_precision = 0
    denum = non_empty_label
    for i in label_list:
        num = precision(result, i)
        if num != -1:
            total_precision += num
        else:
            print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            non_empty_label -= 1
    precision_m = total_precision / denum
    total_precision_after_iteration += precision_m
    print("Precision:", precision_m)

    # Recall
    total_recall = 0
    denum = non_empty_label
    for i in label_list: # labels are 1, 2, 3
        num = recall(result, i)
        if num != -1:
            total_recall += num
        else:
            print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            denum -= 1
    recall_m = total_recall / non_empty_label
    total_recall_after_iteration += recall_m
    print("Recall:", recall_m)

    # F-score
    f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
    total_f_score_after_iteration += f_score
    print("Macro-averaging F-score:", f_score)
    print()

print("Summary")
print("--------------------")
print("Average accuracy after 3 iterations:", total_accuracy_after_iteration / 3)
print("Average precision after 3 iterations:", total_precision_after_iteration / 3)
print("Average recall after 3 iterations:", total_recall_after_iteration / 3)
print("Average F-score after 3 iterations:", total_f_score_after_iteration / 3)

Iteration 1
--------------------
actual      1    2    3    total
--------  ---  ---  ---  -------
1          65   11    1       77
2           6   58    1       65
3           0   11    0       11
total      71   80    2      153

Non-empty labels count: 3
Average accuracy: 0.869281045751634
Precision: 0.546830985915493
Recall: 0.5788211788211788
Macro-averaging F-score: 0.5623715136861966



Iteration 2
--------------------
actual      1    2    3    total
--------  ---  ---  ---  -------
1          68    5    1       74
2          14   35   17       66
3           0    4    9       13
total      82   44   27      153

Non-empty labels count: 3
Average accuracy: 0.8213507625272332
Precision: 0.6526853904902684
Recall: 0.7138432138432137
Macro-averaging F-score: 0.6818957690290486

Iteration 3
--------------------
actual      1    2    3    total
--------  ---  ---  ---  -------
1          72    8    0       80
2           8   54    0       62
3           0   11    0       11
total      80   73    0      153

Non-empty labels count: 3
Average accuracy: 0.8823529411764707
p_prediced is 0 for label 3 . Subtracted 1 from non_empty_label.
Precision: 0.5465753424657535
Recall: 0.885483870967742
Macro-averaging F-score: 0.6759268687803749

Summary
--------------------
Average accuracy after 3 iterations: 0.8576615831517792
Average precision after 3 iterations: 0.5820305729571715


### 2.8 Average result after running three times using Weighted Vote
Run the kNN algorithm again
* k = 5
* Weighted vote
* Numerical columns: cylinders, weight, and acceleration
* No nominal attributes

Find the accuracy and macro-average f-measure

In [28]:
total_accuracy_after_iteration = 0
total_precision_after_iteration = 0
total_recall_after_iteration = 0
total_f_score_after_iteration = 0

for i in range(3):
    # Discretize
    auto_copy = auto.copy()
    mpg_value = auto_copy.get_column_data('mpg')
    mpg_value = sorted(mpg_value)
    bin_width = (mpg_value[-1] - mpg_value[0]) / 3
    cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width]
    discretize(auto_copy, 'mpg', cut_points)

    # Normalize
    normalize(auto_copy, 'cyls')
    normalize(auto_copy, 'disp')
    normalize(auto_copy, 'hp')
    normalize(auto_copy, 'weight')
    normalize(auto_copy, 'accl')

    # Holdout
    train, test = holdout(auto_copy, auto_copy.row_count()//2)

    # kNN
    result = knn_eval(train, test, weighted_vote, 5, 'mpg', ['cyls','weight','accl'], [])
    label_list = result.get_column_data('actual')
    if 'total' in label_list:
        label_list.remove('total')

    print("Iteration", i+1)
    print("--------------------")
    print(result)
    print()

    # Find non-empty labels
    sum_of_actual_labels = result.get_column_data('total')
    non_empty_label = 0
    for i in sum_of_actual_labels:
        if i != 0:
            non_empty_label += 1

    if result[result.row_count() - 1]['actual'] == 'total':
        non_empty_label -= 1

    print("Non-empty labels count:", non_empty_label)

    # Accuracy
    total_accuracy = 0
    for i in label_list: 
        num = accuracy(result, i)
        if num != -1:
            total_accuracy += num
    acc = total_accuracy / non_empty_label
    total_accuracy_after_iteration += acc
    print("Average accuracy:", acc)

    # Precision
    total_precision = 0
    denum = non_empty_label
    for i in label_list:
        num = precision(result, i)
        if num != -1:
            total_precision += num
        else:
            print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            non_empty_label -= 1
    precision_m = total_precision / denum
    total_precision_after_iteration += precision_m
    print("Precision:", precision_m)

    # Recall
    total_recall = 0
    denum = non_empty_label
    for i in label_list: # labels are 1, 2, 3
        num = recall(result, i)
        if num != -1:
            total_recall += num
        else:
            print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            denum -= 1
    recall_m = total_recall / non_empty_label
    total_recall_after_iteration += recall_m
    print("Recall:", recall_m)

    # F-score
    f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
    total_f_score_after_iteration += f_score
    print("Macro-averaging F-score:", f_score)
    print()

print("Summary")
print("--------------------")
print("Average accuracy after 3 iterations:", total_accuracy_after_iteration / 3)
print("Average precision after 3 iterations:", total_precision_after_iteration / 3)
print("Average recall after 3 iterations:", total_recall_after_iteration / 3)
print("Average F-score after 3 iterations:", total_f_score_after_iteration / 3)

Iteration 1
--------------------
actual      1    2    3    total
--------  ---  ---  ---  -------
1          48   28    7       83
2           6    6   48       60
3           0    0   10       10
total      54   34   65      153

Non-empty labels count: 3
Average accuracy: 0.6122004357298475
Precision: 0.4064018769901123
Recall: 0.559437751004016
Macro-averaging F-score: 0.47079565898396

Iteration 2
--------------------
actual      1    2    3    total
--------  ---  ---  ---  -------
1          42   29    4       75
2           3   32   32       67
3           0    1   10       11
total      45   62   46      153

Non-empty labels count: 3
Average accuracy: 0.699346405228758
Precision: 0.5556178899797414
Recall: 0.6489009497964723
Macro-averaging F-score: 0.5986473015212458

Iteration 3
--------------------
actual      1    2    3    total
--------  ---  ---  ---  -------
1          16   59    5       80
2           5   12   47       64
3           0    0    9        9
total      2

### 2.9 Compare 2.6 and 2.7

Between the two, the results are as follow:

|Metrics|2.6|2.7|
|---|---|---|
|Accuracy|0.8257|0.8577|
|Precision|0.6608|0.5820|
|Recall|0.7233|0.7260|
|F-score|0.6907|0.6401|

* The overall accuracy is higher in 2.7 than 2.6, although they are not siginificantly different.
* The precision is higher in 2.6 than 2.7. This is interesting as the gap is quite big. This might be cause by the Holdout method as it does not maintain the distribution.
* The recall is higher in 2.7 than 2.6, though very slightly.
* The F-score is higher in 2.6 than 2.7.

### 2.10 Redo 2.6 and compare the results at two different bin sizes

At bin=4

In [22]:
total_accuracy_after_iteration = 0
total_precision_after_iteration = 0
total_recall_after_iteration = 0
total_f_score_after_iteration = 0

for i in range(3):
    # Discretize
    auto_copy = auto.copy()
    mpg_value = auto_copy.get_column_data('mpg')
    mpg_value = sorted(mpg_value)
    bin_width = (mpg_value[-1] - mpg_value[0]) / 4
    cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width, mpg_value[0] + 3 * bin_width]
    discretize(auto_copy, 'mpg', cut_points)

    # Normalize
    normalize(auto_copy, 'cyls')
    normalize(auto_copy, 'disp')
    normalize(auto_copy, 'hp')
    normalize(auto_copy, 'weight')
    normalize(auto_copy, 'accl')

    # Holdout
    train, test = holdout(auto_copy, auto_copy.row_count()//2)

    # kNN
    result = knn_eval(train, test, majority_vote, 5, 'mpg', ['cyls','weight','accl'], [])
    label_list = result.get_column_data('actual')
    if 'total' in label_list:
        label_list.remove('total')

    print("Iteration", i+1)
    print("--------------------")
    print(result)
    print()

    # Find non-empty labels
    sum_of_actual_labels = result.get_column_data('total')
    non_empty_label = 0
    for i in sum_of_actual_labels:
        if i != 0:
            non_empty_label += 1

    if result[result.row_count() - 1]['actual'] == 'total':
        non_empty_label -= 1

    print("Non-empty labels count:", non_empty_label)

    # Accuracy
    total_accuracy = 0
    for i in label_list: 
        num = accuracy(result, i)
        if num != -1:
            total_accuracy += num
    acc = total_accuracy / non_empty_label
    total_accuracy_after_iteration += acc
    print("Average accuracy:", acc)

    # Precision
    total_precision = 0
    denum = non_empty_label
    for i in label_list:
        num = precision(result, i)
        if num != -1:
            total_precision += num
        else:
            print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            non_empty_label -= 1
    precision_m = total_precision / denum
    total_precision_after_iteration += precision_m
    print("Precision:", precision_m)

    # Recall
    total_recall = 0
    denum = non_empty_label
    for i in label_list: # labels are 1, 2, 3
        num = recall(result, i)
        if num != -1:
            total_recall += num
        else:
            print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            denum -= 1
    recall_m = total_recall / non_empty_label
    total_recall_after_iteration += recall_m
    print("Recall:", recall_m)

    # F-score
    f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
    total_f_score_after_iteration += f_score
    print("Macro-averaging F-score:", f_score)
    print()

print("Summary")
print("--------------------")
print("Average accuracy after 3 iterations:", total_accuracy_after_iteration / 3)
print("Average precision after 3 iterations:", total_precision_after_iteration / 3)
print("Average recall after 3 iterations:", total_recall_after_iteration / 3)
print("Average F-score after 3 iterations:", total_f_score_after_iteration / 3)

Iteration 1
--------------------
actual      1    2    3    4    total
--------  ---  ---  ---  ---  -------
1          25   24    1    0       50
2           0   63    8    0       71
3           0   26    1    0       27
4           0    5    0    0        5
total      25  118   10    0      153

Non-empty labels count: 4
Average accuracy: 0.7908496732026145
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
Precision: 0.40847457627118644
Recall: 0.4747869935663363
Macro-averaging F-score: 0.4391415241846308

Iteration 2
--------------------
actual      1    2    3    4    total
--------  ---  ---  ---  ---  -------
1          46    8    0    0       54
2           7   44   15    0       66
3           0   24    7    0       31
4           0    2    0    0        2
total      53   78   22    0      153

Non-empty labels count: 4
Average accuracy: 0.8169934640522876
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
Precision: 0.43755222764656726
Recall: 0.

At bin=5

In [23]:
total_accuracy_after_iteration = 0
total_precision_after_iteration = 0
total_recall_after_iteration = 0
total_f_score_after_iteration = 0

for i in range(3):
    # Discretize
    auto_copy = auto.copy()
    mpg_value = auto_copy.get_column_data('mpg')
    mpg_value = sorted(mpg_value)
    bin_width = (mpg_value[-1] - mpg_value[0]) / 5
    cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width, mpg_value[0] + 3 * bin_width, mpg_value[0] + 4 * bin_width]
    discretize(auto_copy, 'mpg', cut_points)

    # Normalize
    normalize(auto_copy, 'cyls')
    normalize(auto_copy, 'disp')
    normalize(auto_copy, 'hp')
    normalize(auto_copy, 'weight')
    normalize(auto_copy, 'accl')

    # Holdout
    train, test = holdout(auto_copy, auto_copy.row_count()//2)

    # kNN
    result = knn_eval(train, test, majority_vote, 5, 'mpg', ['cyls','weight','accl'], [])
    label_list = result.get_column_data('actual')
    if 'total' in label_list:
        label_list.remove('total')

    print("Iteration", i+1)
    print("--------------------")
    print(result)
    print()

    # Find non-empty labels
    sum_of_actual_labels = result.get_column_data('total')
    non_empty_label = 0
    for i in sum_of_actual_labels:
        if i != 0:
            non_empty_label += 1

    if result[result.row_count() - 1]['actual'] == 'total':
        non_empty_label -= 1

    print("Non-empty labels count:", non_empty_label)

    # Accuracy
    total_accuracy = 0
    for i in label_list: 
        num = accuracy(result, i)
        if num != -1:
            total_accuracy += num
    acc = total_accuracy / non_empty_label
    total_accuracy_after_iteration += acc
    print("Average accuracy:", acc)

    # Precision
    total_precision = 0
    denum = non_empty_label
    for i in label_list:
        num = precision(result, i)
        if num != -1:
            total_precision += num
        else:
            print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            non_empty_label -= 1
    precision_m = total_precision / denum
    total_precision_after_iteration += precision_m
    print("Precision:", precision_m)

    # Recall
    total_recall = 0
    denum = non_empty_label
    for i in label_list: # labels are 1, 2, 3
        num = recall(result, i)
        if num != -1:
            total_recall += num
        else:
            print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            denum -= 1
    recall_m = total_recall / non_empty_label
    total_recall_after_iteration += recall_m
    print("Recall:", recall_m)

    # F-score
    f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
    total_f_score_after_iteration += f_score
    print("Macro-averaging F-score:", f_score)
    print()

print("Summary")
print("--------------------")
print("Average accuracy after 3 iterations:", total_accuracy_after_iteration / 3)
print("Average precision after 3 iterations:", total_precision_after_iteration / 3)
print("Average recall after 3 iterations:", total_recall_after_iteration / 3)
print("Average F-score after 3 iterations:", total_f_score_after_iteration / 3)

Iteration 1
--------------------
actual      1    2    3    4    5    total
--------  ---  ---  ---  ---  ---  -------
1          22   11    0    0    0       33
2           2   44   10    0    0       56
3           1    1   41    0    0       43
4           0    0   20    0    0       20
5           0    0    1    0    0        1
total      25   56   72    0    0      153

Non-empty labels count: 5
Average accuracy: 0.8797385620915034
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
p_prediced is 0 for label 5 . Subtracted 1 from non_empty_label.
Precision: 0.447031746031746
Recall: 0.8019564414913253
Macro-averaging F-score: 0.5740646578767594

Iteration 2
--------------------
actual      1    2    3    4    5    total
--------  ---  ---  ---  ---  ---  -------
1          25   14    0    0    0       39
2          13   33   11    0    0       57
3           1    3   33    2    0       39
4           0    0   16    0    0       16
5           0    0    2    0    0    

Comparisons

The results are as follow:
|Metrics|bin=4|bin=5|
|---|---|---|
|Accuracy|0.8181|0.8475|
|Precision|0.4586|0.4251|
|Recall|0.6025|0.6468|
|F-score|0.5199|0.5103|

* Accuracy is higher at bin=5 than bin=4.
* Precision is higher at bin=4 than bin=5, though not significantly.
* Recall is higher at bin=5 than bin=4.
* F-score is higher at bin=4 than bin=5, though not significantly.


### 2.11 Redo 2.6 and compare the results at bin=4 and two different k values

At k=10

In [27]:
total_accuracy_after_iteration = 0
total_precision_after_iteration = 0
total_recall_after_iteration = 0
total_f_score_after_iteration = 0

for i in range(3):
    # Discretize
    auto_copy = auto.copy()
    mpg_value = auto_copy.get_column_data('mpg')
    mpg_value = sorted(mpg_value)
    bin_width = (mpg_value[-1] - mpg_value[0]) / 4
    cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width, mpg_value[0] + 3 * bin_width]
    discretize(auto_copy, 'mpg', cut_points)

    # Normalize
    normalize(auto_copy, 'cyls')
    normalize(auto_copy, 'disp')
    normalize(auto_copy, 'hp')
    normalize(auto_copy, 'weight')
    normalize(auto_copy, 'accl')

    # Holdout
    train, test = holdout(auto_copy, auto_copy.row_count()//2)

    # kNN
    result = knn_eval(train, test, majority_vote, 10, 'mpg', ['cyls','weight','accl'], [])
    label_list = result.get_column_data('actual')
    if 'total' in label_list:
        label_list.remove('total')

    print("Iteration", i+1)
    print("--------------------")
    print(result)
    print()

    # Find non-empty labels
    sum_of_actual_labels = result.get_column_data('total')
    non_empty_label = 0
    for i in sum_of_actual_labels:
        if i != 0:
            non_empty_label += 1

    if result[result.row_count() - 1]['actual'] == 'total':
        non_empty_label -= 1

    print("Non-empty labels count:", non_empty_label)

    # Accuracy
    total_accuracy = 0
    for i in label_list: 
        num = accuracy(result, i)
        if num != -1:
            total_accuracy += num
    acc = total_accuracy / non_empty_label
    total_accuracy_after_iteration += acc
    print("Average accuracy:", acc)

    # Precision
    total_precision = 0
    denum = non_empty_label
    for i in label_list:
        num = precision(result, i)
        if num != -1:
            total_precision += num
        else:
            print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            non_empty_label -= 1
    precision_m = total_precision / denum
    total_precision_after_iteration += precision_m
    print("Precision:", precision_m)

    # Recall
    total_recall = 0
    denum = non_empty_label
    for i in label_list: # labels are 1, 2, 3
        num = recall(result, i)
        if num != -1:
            total_recall += num
        else:
            print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            denum -= 1
    recall_m = total_recall / non_empty_label
    total_recall_after_iteration += recall_m
    print("Recall:", recall_m)

    # F-score
    f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
    total_f_score_after_iteration += f_score
    print("Macro-averaging F-score:", f_score)
    print()

print("Summary")
print("--------------------")
print("Average accuracy after 3 iterations:", total_accuracy_after_iteration / 3)
print("Average precision after 3 iterations:", total_precision_after_iteration / 3)
print("Average recall after 3 iterations:", total_recall_after_iteration / 3)
print("Average F-score after 3 iterations:", total_f_score_after_iteration / 3)

Iteration 1
--------------------
actual      1    2    3    4    total
--------  ---  ---  ---  ---  -------
1          50    9    0    0       59
2           9   29   32    0       70
3           0    2   18    0       20
4           0    0    4    0        4
total      59   40   54    0      153

Non-empty labels count: 4
Average accuracy: 0.8169934640522876
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
Precision: 0.47644774011299434
Recall: 0.7205811138014528
Macro-averaging F-score: 0.5736189935875079

Iteration 2
--------------------
actual      1    2    3    4    total
--------  ---  ---  ---  ---  -------
1          38    7    0    0       45
2           3   37   31    0       71
3           0    5   27    0       32
4           0    0    5    0        5
total      41   49   63    0      153

Non-empty labels count: 4
Average accuracy: 0.8333333333333334
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
Precision: 0.5276256844201095
Recall: 0.7

At k=20

In [26]:
total_accuracy_after_iteration = 0
total_precision_after_iteration = 0
total_recall_after_iteration = 0
total_f_score_after_iteration = 0

for i in range(3):
    # Discretize
    auto_copy = auto.copy()
    mpg_value = auto_copy.get_column_data('mpg')
    mpg_value = sorted(mpg_value)
    bin_width = (mpg_value[-1] - mpg_value[0]) / 4
    cut_points = [mpg_value[0] + bin_width, mpg_value[0] + 2 * bin_width, mpg_value[0] + 3 * bin_width]
    discretize(auto_copy, 'mpg', cut_points)

    # Normalize
    normalize(auto_copy, 'cyls')
    normalize(auto_copy, 'disp')
    normalize(auto_copy, 'hp')
    normalize(auto_copy, 'weight')
    normalize(auto_copy, 'accl')

    # Holdout
    train, test = holdout(auto_copy, auto_copy.row_count()//2)

    # kNN
    result = knn_eval(train, test, majority_vote, 20, 'mpg', ['cyls','weight','accl'], [])
    label_list = result.get_column_data('actual')
    if 'total' in label_list:
        label_list.remove('total')

    print("Iteration", i+1)
    print("--------------------")
    print(result)
    print()

    # Find non-empty labels
    sum_of_actual_labels = result.get_column_data('total')
    non_empty_label = 0
    for i in sum_of_actual_labels:
        if i != 0:
            non_empty_label += 1

    if result[result.row_count() - 1]['actual'] == 'total':
        non_empty_label -= 1

    print("Non-empty labels count:", non_empty_label)

    # Accuracy
    total_accuracy = 0
    for i in label_list: 
        num = accuracy(result, i)
        if num != -1:
            total_accuracy += num
    acc = total_accuracy / non_empty_label
    total_accuracy_after_iteration += acc
    print("Average accuracy:", acc)

    # Precision
    total_precision = 0
    denum = non_empty_label
    for i in label_list:
        num = precision(result, i)
        if num != -1:
            total_precision += num
        else:
            print("p_prediced is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            non_empty_label -= 1
    precision_m = total_precision / denum
    total_precision_after_iteration += precision_m
    print("Precision:", precision_m)

    # Recall
    total_recall = 0
    denum = non_empty_label
    for i in label_list: # labels are 1, 2, 3
        num = recall(result, i)
        if num != -1:
            total_recall += num
        else:
            print("p_actual is 0 for label", i, ". Subtracted 1 from non_empty_label.")
            denum -= 1
    recall_m = total_recall / non_empty_label
    total_recall_after_iteration += recall_m
    print("Recall:", recall_m)

    # F-score
    f_score = 2 * precision_m * recall_m / (precision_m + recall_m)
    total_f_score_after_iteration += f_score
    print("Macro-averaging F-score:", f_score)
    print()

print("Summary")
print("--------------------")
print("Average accuracy after 3 iterations:", total_accuracy_after_iteration / 3)
print("Average precision after 3 iterations:", total_precision_after_iteration / 3)
print("Average recall after 3 iterations:", total_recall_after_iteration / 3)
print("Average F-score after 3 iterations:", total_f_score_after_iteration / 3)

Iteration 1
--------------------
actual      1    2    3    4    total
--------  ---  ---  ---  ---  -------
1          39   11    0    0       50
2           6   26   40    0       72
3           0    1   26    0       27
4           0    0    4    0        4
total      45   38   70    0      153

Non-empty labels count: 4
Average accuracy: 0.7973856209150327
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
Precision: 0.4805764411027569
Recall: 0.7013580246913581
Macro-averaging F-score: 0.5703465855335254

Iteration 2
--------------------
actual      1    2    3    4    total
--------  ---  ---  ---  ---  -------
1          40    9    0    0       49
2           7   38   26    0       71
3           0    3   25    0       28
4           0    0    5    0        5
total      47   50   56    0      153

Non-empty labels count: 4
Average accuracy: 0.8366013071895425
p_prediced is 0 for label 4 . Subtracted 1 from non_empty_label.
Precision: 0.5143731003039514
Recall: 0.74

Comparisons

The results are as follow:
|Metrics|k=10|k=20|
|---|---|---|
|Accuracy|0.8366|0.8126|
|Precision|0.5190|0.4402|
|Recall|0.7495|0.7474|
|F-score|0.6132|0.5492|

* Accuracy is higher at k=10 than k=20, though not significantly.
* Precision is higher at k=10 than k=20. The gap is quite big, meaning when k is higher, the precision is lower. This may also suggest that the optimal k value is between 10 and 20.
* Recall is higher at k=10 than k=20, though slightly.
* F-score is higher at k=10 than k=20. This also suggests the same thing as the precision.

## 3. Issues, Challenges, and Observations

### 3.1 Issues and Challenges
* I don't have any significant issues completing this assignment.
* I was stuck at knn_eval() since I did not know what to expect from the returning result. Besides, I did not think about dropping the label column before passing the training set to calculate the Eucledian distance for KNN. Therefore, it took me a while to figure out what to do and debug the code.

### 3.2 Observations
* I notice a higher score when using majority vote over weighted vote. This is interesting as I thought weighted vote would be better since it takes into account the distance between the points. This might be casued by the amount of data is not that sufficient to make a difference. Besides, since I also scaled the scores that were used for the weighted vote, it might have caused the result to be lower.
* For this set of data, I think the optimal equal-width bin size is 3. At bin=3, the accuracy and f-score yield the highest as compared to bin=4 or bin=5. Otherwise, if we use other discretization methods that can preserve the distribution, the result might be better.
* The optimal k value is between 10 and 20. At k=10, the results are better at k=5 and k=20. We can put this into a loop and run, while comparing the results to find the optimal k value.
* The Holdout method is not that good as it does not maintain the distribution. This is interesting as in some extreme cases, it may produce a test set that does not contain every label. Therefore, the result might be biased. We can use other methods such as k-fold cross validation to get a better result.
